lib/active_merchant/billing/gateways/visanet_peru.rb

```diff
@@ -34,25 +34,33 @@ def authorize(amount, payment_method, options={})
         params[:email] = options[:email] || 'unknown@email.com'
         params[:createAlias] = false
 
-        commit("authorize", params)
+        commit("authorize", params, options)
       end
 
       def capture(authorization, options={})
         params = {}
+        options[:id_unico] = parse_auth(authorization)[1]
         add_auth_order_id(params, authorization, options)
-        commit("deposit", params)
+        commit("deposit", params, options)
       end
 
       def void(authorization, options={})
         params = {}
         add_auth_order_id(params, authorization, options)
-        commit("void", params)
+        commit("void", params, options)
       end
 
       def refund(amount, authorization, options={})
+        # Attempt RefundSingleTransaction if unsettled
         params = {}
+        params[:amount] = amount(amount)
         add_auth_order_id(params, authorization, options)
-        commit("cancelDeposit", params)
+        response = commit("cancelDeposit", params, options)
+        return response if response.success? || parse_auth(authorization).length == 1
+
+
+        prepare_refund_data(params, authorization, options)
```

Comment:
Hmm. _Incoming sophomoric opinion expression_: Instead of this method that mutates a bunch of stuff, I think it would be a bit cleaner to have an explicit method for building a `cancelDeposit` transaction and an explicit method for building a `refundSingleTransaction` transaction, and the `refund` method would then call `refundSingleTransaction` if `cancelDeposit` fails. It would also make it easier to call one or the other individually, if we need to do that in the future for some reason. 

Maybe premature optimization, and I don't know if you looked into that already and it wasn't as simple as it sounds, but consider it if not?

Reply:
Definitely will consider. I generally try to touch others code as little as possible, but definitely agree with your opinion.